SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp

```diff
@@ -46,15 +46,43 @@ void USpatialLoadBalanceEnforcer::OnAuthorityIntentComponentUpdated(const Worker
 	}
 }
 
+// This is called whenever this worker becomes authoritative for the ACL component on an entity.
+// It is now this worker's responsibility to make sure that the ACL reflects the current intent,
+// which may have been received before this call.
 void USpatialLoadBalanceEnforcer::AuthorityChanged(const Worker_AuthorityChangeOp& AuthOp)
 {
 	if (AuthOp.component_id == SpatialConstants::ENTITY_ACL_COMPONENT_ID &&
 		AuthOp.authority == WORKER_AUTHORITY_AUTHORITATIVE)
 	{
+		const AuthorityIntent* AuthorityIntentComponent = StaticComponentView->GetComponentData<SpatialGDK::AuthorityIntent>(AuthOp.entity_id);
+		if (AuthorityIntentComponent == nullptr)
+		{
+			// TODO(zoning): There are still some entities being created without an authority intent component.
+			// For example, the Unreal created worker entities don't have one. Even though those won't be able to
+			// transition, we should have the intent component on them for completeness.
+		 	UE_LOG(LogSpatialLoadBalanceEnforcer, Log, TEXT("Requested authority change for entity without AuthorityIntent component. EntityId: %lld"), AuthOp.entity_id);
+			return;
+		}
+
+		const FString* OwningWorkerId = VirtualWorkerTranslator->GetPhysicalWorkerForVirtualWorker(AuthorityIntentComponent->VirtualWorkerId);
+		if (OwningWorkerId != nullptr &&
+			*OwningWorkerId == WorkerId &&
+			StaticComponentView->GetAuthority(AuthOp.entity_id, SpatialConstants::AUTHORITY_INTENT_COMPONENT_ID) == WORKER_AUTHORITY_AUTHORITATIVE)
+		{
+			UE_LOG(LogSpatialLoadBalanceEnforcer, Verbose, TEXT("No need to queue newly authoritative entity %lld because this worker is already authoritative."), AuthOp.entity_id);
+			return;
+		}
 		QueueAclAssignmentRequest(AuthOp.entity_id);
 	}
 }
 
+// QueueAclAssignmentRequest is called from three places.
+// 1) AuthorityIntent change - Intent is not authoritative on this worker - ACL is authoritative on this worker.
+//    (another worker changed the intent, but this worker is responsible for the ACL, so update it.)
+// 2) ACL change - Intent is unknown - ACL just became authoritative on this worker.
```

Comment:
`Intent is unknown` - To me, and this could be incorrect, this implies this worker may not having the `AuthorityIntent `component data but I don't think that can ever be true since `AuthorityIntent` exists on every entity and will be present during the critical section?

Reply:
I forgot we have the critical section stuff in this. +1 I think all components have to be added before we process authority.

Reply:
I phrased the comment badly. It's not that intent isn't known on the worker, it's that this bit of code hasn't checked it yet. The interpretative line on the next line makes that clearer, but I'll rewrite this to make clearer as well.